# IBM Data Science: Capstone Project
## The Capstone Notebook


### Introduction

Before COVID-19 threw all other business strategies or plans into complete disarray, Europe was still getting to grips on how Brexit was going to affect the region.

Now that Britain has chosen to leave the EU, many businesses will relocate their head offices outside of London, the world's former financial centre. With this relocation, hundreds of thousands (if not millions) of people will have to uplift their lives and move to another city.

Based on this [forbes article](https://fortune.com/longform/brexit-amsterdam-the-new-london-europe-companies/) Amsterdam could be the 'new London'. If employees are forced to move from the one city to the other, the big question is: which neighborhoods should these soon-to-be ex-Londoners choose to move to in Amsterdam?

Based on Foursquare data, I aim to cluster Amsterdam and London neighborhoods based on similarity. So if you love your neighborhood in London and want something similar in Amsterdam, this is your go-to guide.

### Data

The data I will be using is from accessing the Foursquare API. I will be exploring nighborhoods in London and Amsterdam, using the **explore** function to get the most common venue categories in each neighborhood in the two cities, and then use this feature to group the neighborhoods into clusters. 


I will use the geopy library to get the latitude and longitude values of London and Amsterdam, then use an unsupervised machine learning clustering algorithm, *k*-means, to do the custering. Lastly, I will use the Folium library to colour code the neighborhoods in Amsterdam and London based on their emerging clusters.

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1. <a href="#item1">Download and Explore Dataset</a>

2. <a href="#item2">Explore Neighborhoods in London and Amsterdam</a>

3. <a href="#item3">Analyze Each Neighborhood</a>

4. <a href="#item4">Cluster Neighborhoods</a>

5. <a href="#item5">Examine Clusters</a>    
</font>
</div>

In [467]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [468]:
import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import GoogleV3 # convert an address into latitude and longitude values

In [469]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

In [470]:
#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## 1. Download and Explore Dataset
### a) Scraping Geo-location Data
### b) Scraping Foursquare Data

First we have to scrape the neighborhood, borough and latitude/longitude data for each city.
* To get the London data I scraped [this](https://en.wikipedia.org/wiki/List_of_areas_of_London) wikipedia page.
* To get the Amsterdam data I scraped [this](https://en.wikipedia.org/wiki/Boroughs_of_Amsterdam) wikipedia page.

Then we have to get the foursquare data

### London

In [471]:
#Scraping the London data
#Reading in the table and setting the header to the first row
df = pd.read_html(r'https://en.wikipedia.org/wiki/List_of_areas_of_London', match = 'Location')
df = pd.DataFrame(df[0])
df.head()

Location                     London borough       Post town  \
0   Abbey Wood              Bexley, Greenwich [7]          LONDON   
1        Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2    Addington                         Croydon[8]         CROYDON   
3   Addiscombe                         Croydon[8]         CROYDON   
4  Albany Park                             Bexley  BEXLEY, SIDCUP   

  Postcode district Dial code OS grid ref  
0               SE2       020    TQ465785  
1            W3, W4       020    TQ205805  
2               CR0       020    TQ375645  
3               CR0       020    TQ345665  
4         DA5, DA14       020    TQ478728

In [472]:
#setting the column labels
df.columns = ['location', 'borough', 'posttown', 'postcode', 'dialcode', 'OSgridref']
df.head()

location                            borough        posttown   postcode  \
0   Abbey Wood              Bexley, Greenwich [7]          LONDON        SE2   
1        Acton  Ealing, Hammersmith and Fulham[8]          LONDON     W3, W4   
2    Addington                         Croydon[8]         CROYDON        CR0   
3   Addiscombe                         Croydon[8]         CROYDON        CR0   
4  Albany Park                             Bexley  BEXLEY, SIDCUP  DA5, DA14   

  dialcode OSgridref  
0      020  TQ465785  
1      020  TQ205805  
2      020  TQ375645  
3      020  TQ345665  
4      020  TQ478728

In [473]:
#dropping unnecessary columns
df.drop(['dialcode', 'OSgridref'], axis =1, inplace = True)
df.columns

Index(['location', 'borough', 'posttown', 'postcode'], dtype='object')

In [474]:
#Only selecting the first postal code and first boroughs listed
first_postcode= df.postcode.str.split(",", expand = True)[0]
df['postcode'] = first_postcode

first_borough= df.borough.str.split(",", expand = True)[0]
df['borough'] = first_borough
df.head()

location     borough        posttown postcode
0   Abbey Wood      Bexley          LONDON      SE2
1        Acton      Ealing          LONDON       W3
2    Addington  Croydon[8]         CROYDON      CR0
3   Addiscombe  Croydon[8]         CROYDON      CR0
4  Albany Park      Bexley  BEXLEY, SIDCUP      DA5

In [475]:
# Cleaning up the borough names to remove the square brackets with regex
df.borough = df.borough.map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df

location  \
0                                        Abbey Wood   
1                                             Acton   
2                                         Addington   
3                                        Addiscombe   
4                                       Albany Park   
5                                  Aldborough Hatch   
6                                           Aldgate   
7                                           Aldwych   
8                                          Alperton   
9                                           Anerley   
10                                            Angel   
11                                        Aperfield   
12                                          Archway   
13                                   Ardleigh Green   
14                                           Arkley   
15                                      Arnos Grove   
16                                           Balham   
17                                         Bankside   
18                                         Barbican   
19                                          Barking   
20                                      Barkingside   
21                                       Barnehurst   
22                                           Barnes   
23                                      Barnes Cray   
24                                      Barnet Gate   
25       Barnet (also Chipping Barnet, High Barnet)   
26                                        Barnsbury   
27                                        Battersea   
28                                        Bayswater   
29                                        Beckenham   
30                                          Beckton   
31                                        Becontree   
32                                  Becontree Heath   
33                                       Beddington   
34                                     Bedford Park   
35                                        Belgravia   
36                                       Bellingham   
37                                          Belmont   
38                                          Belmont   
39                                     Belsize Park   
40                                        Belvedere   
41                                       Bermondsey   
42                                       Berrylands   
43                                    Bethnal Green   
44         Bexley (also Old Bexley, Bexley Village)   
45               Bexleyheath (also Bexley New Town)   
46                                          Bickley   
47                                      Biggin Hill   
48                                         Blackfen   
49                                      Blackfriars   
50                                       Blackheath   
51                        Blackheath Royal Standard   
52                                        Blackwall   
53                                          Blendon   
54                                       Bloomsbury   
55                                       Botany Bay   
56                                     Bounds Green   
57                                              Bow   
58                                       Bowes Park   
59                                        Brentford   
60                                      Brent Cross   
61                                       Brent Park   
62                                        Brimsdown   
63                                          Brixton   
64                                         Brockley   
65                                          Bromley   
66                    Bromley (also Bromley-by-Bow)   
67                                   Bromley Common   
68                                         Brompton   
69                                      Brondesbury   
70                                   Brunswick Park   
71                                      Bulls Cross   
72                          

In [476]:
#only selecting London Neighborhoods and cross referencing that the selection is correct
df = df[df.posttown == "LONDON"]
df.posttown.value_counts()

LONDON    299
Name: posttown, dtype: int64

In [477]:
#dropping the posttown field as it doesnt add value anymore
df.drop(['posttown'], axis =1, inplace = True)
df.reset_index(drop=True, inplace = True)

/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/frame.py:3940: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [478]:
df.head()

location      borough postcode
0  Abbey Wood       Bexley      SE2
1       Acton       Ealing       W3
2     Aldgate         City      EC3
3     Aldwych  Westminster      WC2
4     Anerley      Bromley     SE20

In [479]:
# The code was removed by Watson Studio for sharing.

In [480]:
from geopy.geocoders import GoogleV3 # convert an address into latitude and longitude values

In [481]:
geolocator = GoogleV3(api_key = googlemaps_api_key, domain = 'maps.google.co.uk')
location = geolocator.geocode('SE2, London, United Kingdom')
location.latitude, location.longitude

(51.4860421, 0.1202586)

In [482]:
def get_lat_long_london(postcode):
    # Initialize the Location (lat. and long.) to "None"
    lat_lng_coords = None
    
    # While loop helps to create a continous run until all the location coordinates are geocoded
    while(lat_lng_coords is None):
        address = '{}, London, United Kingdom'.format(postcode)
        geolocator = GoogleV3(api_key = googlemaps_api_key, domain = 'maps.google.co.uk')
        location = geolocator.geocode(address)
        latitude = location.latitude
        longitude = location.longitude
        lat_lng_coords = [latitude, longitude]
    return lat_lng_coords


In [483]:
#testing
get_lat_long_london('SE2')

[51.4860421, 0.1202586]

In [484]:
#Using comprehension to create the data for our additional columns
post_codes = [get_lat_long_london(p) for p in df['postcode'].to_list()]

In [485]:
#creating the dataframe which we will append to the end of our existing dataframe
df_c = pd.DataFrame(post_codes, columns = ['latitude', 'longitude'])
df_c.head()

latitude  longitude
0  51.486042   0.120259
1  51.512075  -0.267572
2  51.515987  -0.078379
3  51.510583  -0.126768
4  51.412571  -0.061397

In [486]:
df_lon = pd.concat([df, df_c], axis=1)
df_lon.head()

location      borough postcode   latitude  longitude
0  Abbey Wood       Bexley      SE2  51.486042   0.120259
1       Acton       Ealing       W3  51.512075  -0.267572
2     Aldgate         City      EC3  51.515987  -0.078379
3     Aldwych  Westminster      WC2  51.510583  -0.126768
4     Anerley      Bromley     SE20  51.412571  -0.061397

### Amsterdam

In [487]:
#Scraping the Amsterdam data
#Reading in the table and setting the header to the first row
df = pd.read_html(r'https://en.wikipedia.org/wiki/Boroughs_of_Amsterdam', match = 'Area')
df = pd.DataFrame(df[0])
df.head()

Borough       Area  Population Population density  \
0      Centrum (Centre)   8.04 km²       86422         13,748/km²   
1         Noord (North)  49.01 km²       94766          2,269/km²   
2  Nieuw-West(New West)  32.38 km²      151677          4,478/km²   
3           Oost (East)  30.56 km²      135767          7,635/km²   
4                  West   9.89 km²      143842         15,252/km²   

   Location (in green)                                     Neighbourhoods  
0                  NaN  Binnenstad, Grachtengordel, Haarlemmerbuurt, J...  
1                  NaN  Banne Buiksloot, Buiksloot, Buikslotermeer, Fl...  
2                  NaN  Geuzenveld, Nieuw Sloten, Oostoever, Osdorp, O...  
3                  NaN  IJburg, Indische Buurt, Eastern Docklands, Oud...  
4                  NaN  Frederik Hendrikbuurt, Houthaven, Spaarndammer...

In [488]:
#renaming the columns fo ease of use
df.columns = ['borough', 'area', 'population', 'populationdensity', 'location', 'neighborhoods']
df.head()

borough       area  population populationdensity  location  \
0      Centrum (Centre)   8.04 km²       86422        13,748/km²       NaN   
1         Noord (North)  49.01 km²       94766         2,269/km²       NaN   
2  Nieuw-West(New West)  32.38 km²      151677         4,478/km²       NaN   
3           Oost (East)  30.56 km²      135767         7,635/km²       NaN   
4                  West   9.89 km²      143842        15,252/km²       NaN   

                                       neighborhoods  
0  Binnenstad, Grachtengordel, Haarlemmerbuurt, J...  
1  Banne Buiksloot, Buiksloot, Buikslotermeer, Fl...  
2  Geuzenveld, Nieuw Sloten, Oostoever, Osdorp, O...  
3  IJburg, Indische Buurt, Eastern Docklands, Oud...  
4  Frederik Hendrikbuurt, Houthaven, Spaarndammer...

In [489]:
#dropping columns that I don't need
df = df[['borough','neighborhoods']]
df.head()

borough                                      neighborhoods
0      Centrum (Centre)  Binnenstad, Grachtengordel, Haarlemmerbuurt, J...
1         Noord (North)  Banne Buiksloot, Buiksloot, Buikslotermeer, Fl...
2  Nieuw-West(New West)  Geuzenveld, Nieuw Sloten, Oostoever, Osdorp, O...
3           Oost (East)  IJburg, Indische Buurt, Eastern Docklands, Oud...
4                  West  Frederik Hendrikbuurt, Houthaven, Spaarndammer...

In [490]:
#We now need to rearrange the table so that we only have a neighborhood and borough column.
#We start by splitting the neighborhood column into many columns, and add the 'neighborhood' prefix for ease of use
df_split= df.neighborhoods.str.split(",", expand = True)
df_split = df_split.add_prefix('Neighborhood_')
df_split.head()

Neighborhood_0   Neighborhood_1      Neighborhood_2  \
0             Binnenstad   Grachtengordel     Haarlemmerbuurt   
1        Banne Buiksloot        Buiksloot      Buikslotermeer   
2             Geuzenveld     Nieuw Sloten           Oostoever   
3                 IJburg   Indische Buurt   Eastern Docklands   
4  Frederik Hendrikbuurt        Houthaven   Spaarndammerbuurt   

       Neighborhood_3    Neighborhood_4 Neighborhood_5 Neighborhood_6  \
0          Jodenbuurt           Jordaan       Kadijken        Lastage   
1           Floradorp          Kadoelen      Molenwijk     Nieuwendam   
2              Osdorp   Overtoomse Veld         Sloten     Slotermeer   
3            Oud-Oost   Watergraafsmeer           None           None   
4   Staatsliedenbuurt    Zeeheldenbuurt     Westerpark    Kinkerbuurt   

                         Neighborhood_7        Neighborhood_8  Neighborhood_9  \
0                      Oosterdokseiland   Oostelijke Eilanden        Plantage   
1   Nieuwendammerdijk en Buiksloterdijk         Oostzanerwerf       Overhoeks   
2                           Slotervaart                  None            None   
3                                  None                  None            None   
4                         Overtoombuurt           De Baarsjes   Bos en Lommer   

        Neighborhood_10     Neighborhood_11       Neighborhood_12  \
0             Rapenburg           Uilenburg   Westelijke Eilanden   
1   Tuindorp Nieuwendam   Tuindorp Oostzaan                  None   
2                  None                None                  None   
3                  None                None                  None   
4         Kolenkitbuurt            Landlust            Sloterdijk   

   Neighborhood_13  
0   Weteringschans  
1             None  
2             None  
3             None  
4             None

In [491]:
#we then join the dataframes back together again and drop the repetitive 'neighborhoods' columns
df_new = pd.concat([df, df_split], axis = 1)
df_new.drop(['neighborhoods'], axis = 1, inplace = True)
df_new

borough         Neighborhood_0   Neighborhood_1  \
0         Centrum (Centre)             Binnenstad   Grachtengordel   
1            Noord (North)        Banne Buiksloot        Buiksloot   
2     Nieuw-West(New West)             Geuzenveld     Nieuw Sloten   
3              Oost (East)                 IJburg   Indische Buurt   
4                     West  Frederik Hendrikbuurt        Houthaven   
5  Westpoort(West Gateway)              Westpoort             None   
6             Zuid (South)            Apollobuurt    Buitenveldert   
7      Zuidoost(Southeast)            Bijlmermeer     Venserpolder   

         Neighborhood_2      Neighborhood_3    Neighborhood_4  \
0       Haarlemmerbuurt          Jodenbuurt           Jordaan   
1        Buikslotermeer           Floradorp          Kadoelen   
2             Oostoever              Osdorp   Overtoomse Veld   
3     Eastern Docklands            Oud-Oost   Watergraafsmeer   
4     Spaarndammerbuurt   Staatsliedenbuurt    Zeeheldenbuurt   
5                  None                None              None   
6   Hoofddorppleinbuurt      Museumkwartier           De Pijp   
7            Gaasperdam            Driemond              None   

        Neighborhood_5  Neighborhood_6                        Neighborhood_7  \
0             Kadijken         Lastage                      Oosterdokseiland   
1            Molenwijk      Nieuwendam   Nieuwendammerdijk en Buiksloterdijk   
2               Sloten      Slotermeer                           Slotervaart   
3                 None            None                                  None   
4           Westerpark     Kinkerbuurt                         Overtoombuurt   
5                 None            None                                  None   
6   Prinses Irenebuurt   Rivierenbuurt                         Schinkelbuurt   
7                 None            None                                  None   

         Neighborhood_8  Neighborhood_9       Neighborhood_10  \
0   Oostelijke Eilanden        Plantage             Rapenburg   
1         Oostzanerwerf       Overhoeks   Tuindorp Nieuwendam   
2                  None            None                  None   
3                  None            None                  None   
4           De Baarsjes   Bos en Lommer         Kolenkitbuurt   
5                  None            None                  None   
6          Stadionbuurt      Vondelpark           Willemspark   
7                  None            None                  None   

      Neighborhood_11       Neighborhood_12  Neighborhood_13  
0           Uilenburg   Westelijke Eilanden   Weteringschans  
1   Tuindorp Oostzaan                  None             None  
2                None                  None             None  
3                None                  None             None  
4            Landlust            Sloterdijk             None  
5                None                  None             None  
6              Zuidas                  None             None  
7                None                  None             None

In [492]:
#we then use melt to get our two columns
df_new.set_index('borough', inplace =  True)
df_new.head()
#df_new.pivot(index = 'borough', columns = df.columns[1:])

Neighborhood_0   Neighborhood_1  \
borough                                                        
Centrum (Centre)                 Binnenstad   Grachtengordel   
Noord (North)               Banne Buiksloot        Buiksloot   
Nieuw-West(New West)             Geuzenveld     Nieuw Sloten   
Oost (East)                          IJburg   Indische Buurt   
West                  Frederik Hendrikbuurt        Houthaven   

                          Neighborhood_2      Neighborhood_3  \
borough                                                        
Centrum (Centre)         Haarlemmerbuurt          Jodenbuurt   
Noord (North)             Buikslotermeer           Floradorp   
Nieuw-West(New West)           Oostoever              Osdorp   
Oost (East)            Eastern Docklands            Oud-Oost   
West                   Spaarndammerbuurt   Staatsliedenbuurt   

                        Neighborhood_4 Neighborhood_5 Neighborhood_6  \
borough                                                                
Centrum (Centre)               Jordaan       Kadijken        Lastage   
Noord (North)                 Kadoelen      Molenwijk     Nieuwendam   
Nieuw-West(New West)   Overtoomse Veld         Sloten     Slotermeer   
Oost (East)            Watergraafsmeer           None           None   
West                    Zeeheldenbuurt     Westerpark    Kinkerbuurt   

                                            Neighborhood_7  \
borough                                                      
Centrum (Centre)                          Oosterdokseiland   
Noord (North)          Nieuwendammerdijk en Buiksloterdijk   
Nieuw-West(New West)                           Slotervaart   
Oost (East)                                           None   
West                                         Overtoombuurt   

                            Neighborhood_8  Neighborhood_9  \
borough                                                      
Centrum (Centre)       Oostelijke Eilanden        Plantage   
Noord (North)                Oostzanerwerf       Overhoeks   
Nieuw-West(New West)                  None            None   
Oost (East)                           None            None   
West                           De Baarsjes   Bos en Lommer   

                           Neighborhood_10     Neighborhood_11  \
borough                                                          
Centrum (Centre)                 Rapenburg           Uilenburg   
Noord (North)          Tuindorp Nieuwendam   Tuindorp Oostzaan   
Nieuw-West(New West)                  None                None   
Oost (East)                           None                None   
West                         Kolenkitbuurt            Landlust   

                           Neighborhood_12  Neighborhood_13  
borough                                                      
Centrum (Centre)       Westelijke Eilanden   Weteringschans  
Noord (North)                         None             None  
Nieuw-West(New West)                  None             None  
Oost (East)                           None             None  
West                            Sloterdijk             None

In [493]:
#Cleaning up the dataframe
df = pd.DataFrame(df_new.stack(level=0))
df.index = df.index.get_level_values('borough')
df.reset_index(drop =False, inplace =True)
df.rename(columns = {0:'neighborhood'}, inplace = True)
df.head()

borough      neighborhood
0  Centrum (Centre)        Binnenstad
1  Centrum (Centre)    Grachtengordel
2  Centrum (Centre)   Haarlemmerbuurt
3  Centrum (Centre)        Jodenbuurt
4  Centrum (Centre)           Jordaan

In [494]:
#we now need to remove the english translation within the brackets by using regex
df['borough'] = df['borough'].str.replace(r'\([^()]*\)', '')
df.head()

borough      neighborhood
0  Centrum         Binnenstad
1  Centrum     Grachtengordel
2  Centrum    Haarlemmerbuurt
3  Centrum         Jodenbuurt
4  Centrum            Jordaan

In [495]:
#Now using the get_lat_lng function we used earlier on the amsterdam data, we have to change the domain and address string:
geolocator = GoogleV3(api_key = googlemaps_api_key, domain = 'maps.google.co.uk')

In [496]:
def get_lat_long_amsterdam(neighborhood):
    # Initialize the Location (lat. and long.) to "None"
    lat_lng_coords = None
    
    # While loop helps to create a continous run until all the location coordinates are geocoded
    while(lat_lng_coords is None):
        address = '{}, Amsterdam, Nederland'.format(neighborhood)
        geolocator = GoogleV3(api_key = googlemaps_api_key, domain = 'maps.google.co.uk')
        location = geolocator.geocode(address)
        latitude = location.latitude
        longitude = location.longitude
        lat_lng_coords = [latitude, longitude]
    return lat_lng_coords


In [499]:
#checking that it works
location = geolocator.geocode('Grachtengordel, Amsterdam, Nederland')
location.latitude, location.longitude
print (location.latitude, location.longitude)
print('It works!')

52.3655774 4.8893609
It works!


In [500]:
#Using comprehension to create the data for our additional columns
post_codes = [get_lat_long_amsterdam(n) for n in df['neighborhood'].to_list()]
post_codes

[[52.3733862, 4.8940639],
 [52.3655774, 4.8893609],
 [52.3842542, 4.8893437],
 [52.36826319999999, 4.904020099999999],
 [52.3739315, 4.880875899999999],
 [52.36955220000001, 4.9146351],
 [52.3754477, 4.9023199],
 [52.375737, 4.9100857],
 [52.3695052, 4.9232739],
 [52.3672287, 4.912542999999999],
 [52.3711467, 4.9087828],
 [52.37013710000001, 4.9026176],
 [52.3855276, 4.8892088],
 [52.359706, 4.8908357],
 [52.4057469, 4.9180914],
 [52.40629449999999, 4.9179696],
 [52.4011401, 4.9358496],
 [52.3989074, 4.9148681],
 [52.418266, 4.9063234],
 [52.4200381, 4.8875712],
 [52.390564, 4.946151299999999],
 [52.3894466, 4.940962],
 [52.4216359, 4.8884659],
 [52.38898450000001, 4.900574199999999],
 [52.390564, 4.946151299999999],
 [52.4119909, 4.8932176],
 [52.3785614, 4.8003825],
 [52.3447535, 4.811520799999999],
 [52.3673227, 4.828265099999999],
 [52.3599222, 4.7891092],
 [52.3641277, 4.8404649],
 [52.3385948, 4.7983068],
 [52.3824555, 4.8063614],
 [52.3562134, 4.827257299999999],
 [52.355, 4.997

In [501]:
#creating the dataframe which we will append to the end of our existing dataframe
df_am_lat_lng = pd.DataFrame(post_codes, columns = ['latitude', 'longitude'])
df_am_lat_lng.head()

latitude  longitude
0  52.373386   4.894064
1  52.365577   4.889361
2  52.384254   4.889344
3  52.368263   4.904020
4  52.373931   4.880876

In [502]:
df_am = pd.concat([df, df_am_lat_lng], axis=1)
df_am.head()

borough      neighborhood   latitude  longitude
0  Centrum         Binnenstad  52.373386   4.894064
1  Centrum     Grachtengordel  52.365577   4.889361
2  Centrum    Haarlemmerbuurt  52.384254   4.889344
3  Centrum         Jodenbuurt  52.368263   4.904020
4  Centrum            Jordaan  52.373931   4.880876

### Now concatenating the two dataframes by row (on top of one another), we get the final dataframe

In [503]:
df_lon.head()

location      borough postcode   latitude  longitude
0  Abbey Wood       Bexley      SE2  51.486042   0.120259
1       Acton       Ealing       W3  51.512075  -0.267572
2     Aldgate         City      EC3  51.515987  -0.078379
3     Aldwych  Westminster      WC2  51.510583  -0.126768
4     Anerley      Bromley     SE20  51.412571  -0.061397

In [504]:
df_am.head()

borough      neighborhood   latitude  longitude
0  Centrum         Binnenstad  52.373386   4.894064
1  Centrum     Grachtengordel  52.365577   4.889361
2  Centrum    Haarlemmerbuurt  52.384254   4.889344
3  Centrum         Jodenbuurt  52.368263   4.904020
4  Centrum            Jordaan  52.373931   4.880876

In [505]:
df_am = df_am[['neighborhood', 'borough', 'latitude', 'longitude']]
df_am['city'] = 'Amsterdam'
df_am.head()

neighborhood   borough   latitude  longitude       city
0        Binnenstad  Centrum   52.373386   4.894064  Amsterdam
1    Grachtengordel  Centrum   52.365577   4.889361  Amsterdam
2   Haarlemmerbuurt  Centrum   52.384254   4.889344  Amsterdam
3        Jodenbuurt  Centrum   52.368263   4.904020  Amsterdam
4           Jordaan  Centrum   52.373931   4.880876  Amsterdam

In [506]:
#Because we dont have the post codes for Amsterdam- and given that we actually don't need them , we can drop it before we concatenate.
df_lon.drop(['postcode'], axis = 1, inplace =True)

In [507]:
df_lon.rename({'location': 'neighborhood'}, axis = 1, inplace = True)
df_lon['city'] = 'london'

In [508]:
df_am_lon = pd.concat([df_lon, df_am])
df_am_lon.reset_index(drop =True, inplace =True)
df_am_lon

neighborhood  \
0                                        Abbey Wood   
1                                             Acton   
2                                           Aldgate   
3                                           Aldwych   
4                                           Anerley   
5                                             Angel   
6                                           Archway   
7                                       Arnos Grove   
8                                            Balham   
9                                          Bankside   
10                                         Barbican   
11                                           Barnes   
12                                        Barnsbury   
13                                        Battersea   
14                                        Bayswater   
15                                     Bedford Park   
16                                        Belgravia   
17                                       Bellingham   
18                                     Belsize Park   
19                                       Bermondsey   
20                                    Bethnal Green   
21                                      Blackfriars   
22                                       Blackheath   
23                        Blackheath Royal Standard   
24                                        Blackwall   
25                                       Bloomsbury   
26                                     Bounds Green   
27                                              Bow   
28                                       Bowes Park   
29                                      Brent Cross   
30                                       Brent Park   
31                                          Brixton   
32                                         Brockley   
33                    Bromley (also Bromley-by-Bow)   
34                                         Brompton   
35                                      Brondesbury   
36                                   Brunswick Park   
37                                   Burroughs, The   
38                                       Camberwell   
39                                  Cambridge Heath   
40                                      Camden Town   
41                                     Canary Wharf   
42                                        Cann Hall   
43                                     Canning Town   
44                                        Canonbury   
45                                        Castelnau   
46                                          Catford   
47                                       Chalk Farm   
48                                    Charing Cross   
49                                         Charlton   
50                                          Chelsea   
51                                      Childs Hill   
52                                        Chinatown   
53                                        Chinbrook   
54                                        Chingford   
55                                         Chiswick   
56                                       Church End   
57                                       Church End   
58                                          Clapham   
59                                      Clerkenwell   
60                                        Colindale   
61                                    Colliers Wood   
62                                     Colney Hatch   
63                                    Covent Garden   
64                                      Cricklewood   
65                                     Crofton Park   
66                                        Crossness   
67                                       Crouch End   
68                                   Crystal Palace   
69                                      Cubitt Town   
70                                     Custom House   
71                                          Dalston   
72                      

In [509]:
# Having a look at the two cities using a folium map
#Looking at Amsterdam first
location = geolocator.geocode( 'Amsterdam, Nederland')
lat_am = location.latitude
lng_am = location.longitude
print (lat_am, lng_am)

52.3666969 4.8945398


In [510]:
# create map of Amsterdam using latitude and longitude values
map_am= folium.Map(location=[lat_am, lng_am], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_am['latitude'], df_am['longitude'], df_am['borough'], df_am['neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_am)  
    
map_am

In [511]:
# Having a look at the two cities using a folium map
#Looking at London
location = geolocator.geocode( 'London, United Kingdom')
lat_lon = location.latitude
lng_lon = location.longitude
print (lat_lon, lng_lon)

51.5073509 -0.1277583


In [512]:
# create map of Amsterdam using latitude and longitude values
map_lon= folium.Map(location=[lat_lon, lng_lon], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_lon['latitude'], df_lon['longitude'], df_lon['borough'], df_lon['neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_lon)  
    
map_lon

In [513]:
df_am_lon.shape

(368, 5)

In [514]:
df_am_lon.borough.nunique()

41

In [515]:
df_am_lon.groupby(by = ['city']).count()['neighborhood']

city
Amsterdam     69
london       299
Name: neighborhood, dtype: int64

In [516]:
a = df_am_lon.groupby(by = ['city'])['borough'].value_counts()

### Findings from our initial neighborhood data

We now have a final dataframe with  368 neighborhoods across both cities.
Given the size of London compared to Amsterdam, it's no surprise that London has over 3 times as many neighborhoods.
* London has 299 Neighborhoods
* Amsterdam has 69 Neighborhoods

In total, there are 41 unique Boroughs between the two cities.
* The Borough with the most neighborhoods in Amsterdam is Centrum with 14 Neighborhoods.
* The Borough with the most neighborhoods in London is Barnet with 26 Neighborhoods.

If we take a look at each folium map, we can see that London as a geographic region is much bigger than Amsterdam.

## 3. Clustering Neighborhoods using FourSquare <a class="anchor" id="third-bullet"></a>

In [517]:
import requests # library to handle requests
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [518]:
# The code was removed by Watson Studio for sharing.

In [519]:
#Exploring the first neighborhood in our dataset- 'Abbey Wood'

df_am_lon.loc[0, 'neighborhood']

'Abbey Wood'

#### Now, let's get the top 50 venues that are in Regent Park within a radius of 500 meters.

In [520]:
neighborhood_name = df_am_lon.loc[0, 'neighborhood']
neighborhood_latitude = df_am_lon.loc[0, 'latitude']
neighborhood_longitude = df_am_lon.loc[0, 'longitude']

LIMIT = 50 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=HSNVL3XPEP1ZEFL5PKKJDKHOFMSNBCM5WQ0JJL51RY5AV3GG&client_secret=GGRC30CTORKYNVWCDJOTTTUIPUJYYPBSBNEABBWXU13OJCOR&v=20180605&ll=51.4860421,0.1202586&radius=500&limit=50'

In [521]:
#Sending the GET request to examine the results
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e94bb9a5fb726001bc4bb2a'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Abbey Wood',
  'headerFullLocation': 'Abbey Wood, London',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 3,
  'suggestedBounds': {'ne': {'lat': 51.4905421045, 'lng': 0.12747164715160642},
   'sw': {'lat': 51.4815420955, 'lng': 0.11304555284839356}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '52ff5c17498ea6224a14a31f',
       'name': 'Greggs',
       'location': {'address': '7 Wilton Rd',
        'lat': 51.490164,
        'lng': 0.121305,
        'labeledLatLngs': [{'label': 'display',
          'lat': 51.490164,


In [522]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [523]:
#Cleaning Json and putting into pandas DF structure
venues = results['response']['groups'][0]['items']
venues

[{'reasons': {'count': 0,
   'items': [{'summary': 'This spot is popular',
     'type': 'general',
     'reasonName': 'globalInteractionReason'}]},
  'venue': {'id': '52ff5c17498ea6224a14a31f',
   'name': 'Greggs',
   'location': {'address': '7 Wilton Rd',
    'lat': 51.490164,
    'lng': 0.121305,
    'labeledLatLngs': [{'label': 'display',
      'lat': 51.490164,
      'lng': 0.121305}],
    'distance': 464,
    'postalCode': 'SE2 9RH',
    'cc': 'GB',
    'city': 'Abbey Wood',
    'state': 'Greater London',
    'country': 'United Kingdom',
    'formattedAddress': ['7 Wilton Rd',
     'Abbey Wood',
     'Greater London',
     'SE2 9RH',
     'United Kingdom']},
   'categories': [{'id': '4bf58dd8d48988d16a941735',
     'name': 'Bakery',
     'pluralName': 'Bakeries',
     'shortName': 'Bakery',
     'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/bakery_',
      'suffix': '.png'},
     'primary': True}],
   'photos': {'count': 0, 'groups': []}},
  'referralId': 'e-0-52

In [524]:
nearby_venues = json_normalize(venues) # flatten JSON
nearby_venues.head()

reasons.count                                      reasons.items  \
0              0  [{'summary': 'This spot is popular', 'type': '...   
1              0  [{'summary': 'This spot is popular', 'type': '...   
2              0  [{'summary': 'This spot is popular', 'type': '...   

                       referralId  \
0  e-0-52ff5c17498ea6224a14a31f-0   
1  e-0-4d826745bb2c5481b93be3ab-1   
2  e-0-5191ffc5498e4913b76a852f-2   

                                    venue.categories  \
0  [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...   
1  [{'id': '4bf58dd8d48988d1e4941735', 'name': 'C...   
2  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...   

                   venue.id venue.location.address venue.location.cc  \
0  52ff5c17498ea6224a14a31f            7 Wilton Rd                GB   
1  4d826745bb2c5481b93be3ab                    NaN                GB   
2  5191ffc5498e4913b76a852f      183 Abbey Wood Rd                GB   

  venue.location.city venue.location.country  venue.location.distance  \
0          Abbey Wood         United Kingdom                      464   
1                 NaN         United Kingdom                       62   
2                 NaN         United Kingdom                      415   

                     venue.location.formattedAddress  \
0  [7 Wilton Rd, Abbey Wood, Greater London, SE2 ...   
1                                   [United Kingdom]   
2       [183 Abbey Wood Rd, SE2 9DZ, United Kingdom]   

                       venue.location.labeledLatLngs  venue.location.lat  \
0  [{'label': 'display', 'lat': 51.490164, 'lng':...           51.490164   
1  [{'label': 'display', 'lat': 51.48550163169358...           51.485502   
2  [{'label': 'display', 'lat': 51.48975372314453...           51.489754   

   venue.location.lng venue.location.postalCode venue.location.state  \
0            0.121305                   SE2 9RH       Greater London   
1            0.120014                       NaN                  NaN   
2            0.120822                   SE2 9DZ                  NaN   

                venue.name  venue.photos.count venue.photos.groups  
0                   Greggs                   0                  []  
1  Abbey Wood Caravan Club                   0                  []  
2               Abbey Cafe                   0                  []

In [525]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]
nearby_venues.head()

venue.name                                   venue.categories  \
0                   Greggs  [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...   
1  Abbey Wood Caravan Club  [{'id': '4bf58dd8d48988d1e4941735', 'name': 'C...   
2               Abbey Cafe  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...   

   venue.location.lat  venue.location.lng  
0           51.490164            0.121305  
1           51.485502            0.120014  
2           51.489754            0.120822

In [526]:
# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)
nearby_venues.head()

venue.name venue.categories  venue.location.lat  \
0                   Greggs           Bakery           51.490164   
1  Abbey Wood Caravan Club       Campground           51.485502   
2               Abbey Cafe             Café           51.489754   

   venue.location.lng  
0            0.121305  
1            0.120014  
2            0.120822

In [527]:
# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]
nearby_venues.columns

Index(['name', 'categories', 'lat', 'lng'], dtype='object')

In [528]:
nearby_venues.head()

name  categories        lat       lng
0                   Greggs      Bakery  51.490164  0.121305
1  Abbey Wood Caravan Club  Campground  51.485502  0.120014
2               Abbey Cafe        Café  51.489754  0.120822

In [529]:
#Creating a function to repeat the same for all neighborhoods in our London and Amsterdam dataset
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [530]:
am_lon_venues = getNearbyVenues(names=df_am_lon['neighborhood'],
                                   latitudes=df_am_lon['latitude'],
                                   longitudes=df_am_lon['longitude']
                                  )


Abbey Wood
Acton
Aldgate
Aldwych
Anerley
Angel
Archway
Arnos Grove
Balham
Bankside
Barbican
Barnes
Barnsbury
Battersea
Bayswater
Bedford Park
Belgravia
Bellingham
Belsize Park
Bermondsey
Bethnal Green
Blackfriars
Blackheath
Blackheath Royal Standard
Blackwall
Bloomsbury
Bounds Green
Bow
Bowes Park
Brent Cross
Brent Park
Brixton
Brockley
Bromley (also Bromley-by-Bow)
Brompton
Brondesbury
Brunswick Park
Burroughs, The
Camberwell
Cambridge Heath
Camden Town
Canary Wharf
Cann Hall
Canning Town
Canonbury
Castelnau
Catford
Chalk Farm
Charing Cross
Charlton
Chelsea
Childs Hill
Chinatown
Chinbrook
Chingford
Chiswick
Church End
Church End
Clapham
Clerkenwell
Colindale
Colliers Wood
Colney Hatch
Covent Garden
Cricklewood
Crofton Park
Crossness
Crouch End
Crystal Palace
Cubitt Town
Custom House
Dalston
Dartford
De Beauvoir Town
Denmark Hill
Deptford
Dollis Hill
Dulwich
Ealing
Earls Court
Earlsfield
East Dulwich
East Finchley
East Ham
East Sheen
Edmonton
Elephant and Castle
Eltham
Farringdon
Finch

In [531]:
#Sizing and getting a view of the data
print(am_lon_venues.shape)
am_lon_venues.head()

(9257, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0   Abbey Wood              51.486042                0.120259   
1   Abbey Wood              51.486042                0.120259   
2   Abbey Wood              51.486042                0.120259   
3        Acton              51.512075               -0.267572   
4        Acton              51.512075               -0.267572   

                     Venue  Venue Latitude  Venue Longitude Venue Category  
0                   Greggs       51.490164         0.121305         Bakery  
1  Abbey Wood Caravan Club       51.485502         0.120014     Campground  
2               Abbey Cafe       51.489754         0.120822           Café  
3                  Bake Me       51.508452        -0.268543       Creperie  
4        The Station House       51.508877        -0.263076            Pub

In [532]:
#Seeing how many venues we get per neighborhood
am_lon_venues.groupby('Neighborhood').count().head()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                            
 Bos en Lommer                      38                      38     38   
 Buiksloot                          10                      10     10   
 Buikslotermeer                     31                      31     31   
 Buitenveldert                      19                      19     19   
 De Baarsjes                        44                      44     44   

                 Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                      
 Bos en Lommer               38               38              38  
 Buiksloot                   10               10              10  
 Buikslotermeer              31               31              31  
 Buitenveldert               19               19              19  
 De Baarsjes                 44               44              44

#### Analysing each neighborhood

In [533]:
# one hot encoding
am_lon_onehot = pd.get_dummies(am_lon_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
am_lon_onehot['Neighborhood'] = am_lon_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [am_lon_onehot.columns[-1]] + list(am_lon_onehot.columns[:-1])
am_lon_onehot = am_lon_onehot[fixed_columns]

am_lon_onehot.head()

Zoo Exhibit  Accessories Store  Afghan Restaurant  African Restaurant  \
0            0                  0                  0                   0   
1            0                  0                  0                   0   
2            0                  0                  0                   0   
3            0                  0                  0                   0   
4            0                  0                  0                   0   

   American Restaurant  Antique Shop  Aquarium  Arcade  Arepa Restaurant  \
0                    0             0         0       0                 0   
1                    0             0         0       0                 0   
2                    0             0         0       0                 0   
3                    0             0         0       0                 0   
4                    0             0         0       0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Arts & Entertainment  Asian Restaurant  Athletics & Sports  \
0                     0                 0                   0   
1                     0                 0                   0   
2                     0                 0                   0   
3                     0                 0                   0   
4                     0                 0                   0   

   Australian Restaurant  Austrian Restaurant  Auto Garage  BBQ Joint  \
0                      0                    0            0          0   
1                      0                    0            0          0   
2                      0                    0            0          0   
3                      0                    0            0          0   
4                      0                    0            0          0   

   Baby Store  Bagel Shop  Bakery  Bar  Baseball Field  Beach  \
0           0           0       1    0               0      0   
1           0           0       0    0               0      0   
2           0           0       0    0               0      0   
3           0           0       0    0               0      0   
4           0           0       0    0               0      0   

   Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  Belgian Restaurant  \
0                0         0            0           0                   0   
1                0         0            0           0                   0   
2                0         0            0           0                   0   
3                0         0            0           0                   0   
4                0         0            0           0                   0   

   Betting Shop  Bike Rental / Bike Share  Bike Shop  Bistro  Boarding House  \
0             0                         0          0       0               0   
1             0                         0          0       0               0   
2             0                         0          0       0               0   
3             0                         0          0       0               0   
4             0                         0          0       0               0   

   Boat or Ferry  Bookstore  Botanical Garden  Boutique  Bowling Alley  \
0              0          0                 0         0              0   
1              0          0                 0         0              0   
2              0          0                 0         0              0   
3              0          0                 0         0              0   
4              0          0                 0         0              0   

   Boxing Gym  Brasserie  Br

In [534]:
#Grouping the rows by neighborhood and taking the mean of the frequency of the occurance
am_lon_grouped = am_lon_onehot.groupby('Neighborhood').mean().reset_index()
am_lon_grouped.head()

Neighborhood  Zoo Exhibit  Accessories Store  Afghan Restaurant  \
0    Bos en Lommer          0.0                0.0                0.0   
1        Buiksloot          0.0                0.0                0.0   
2   Buikslotermeer          0.0                0.0                0.0   
3    Buitenveldert          0.0                0.0                0.0   
4      De Baarsjes          0.0                0.0                0.0   

   African Restaurant  American Restaurant  Antique Shop  Aquarium  Arcade  \
0                 0.0                  0.0           0.0       0.0     0.0   
1                 0.0                  0.0           0.0       0.0     0.0   
2                 0.0                  0.0           0.0       0.0     0.0   
3                 0.0                  0.0           0.0       0.0     0.0   
4                 0.0                  0.0           0.0       0.0     0.0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0               0.0                     0.0          0.0         0.0   
1               0.0                     0.0          0.0         0.0   
2               0.0                     0.0          0.0         0.0   
3               0.0                     0.0          0.0         0.0   
4               0.0                     0.0          0.0         0.0   

   Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0                  0.0                   0.0          0.000000   
1                  0.0                   0.0          0.000000   
2                  0.0                   0.0          0.032258   
3                  0.0                   0.0          0.000000   
4                  0.0                   0.0          0.000000   

   Athletics & Sports  Australian Restaurant  Austrian Restaurant  \
0                 0.0                    0.0                  0.0   
1                 0.0                    0.0                  0.0   
2                 0.0                    0.0                  0.0   
3                 0.0                    0.0                  0.0   
4                 0.0                    0.0                  0.0   

   Auto Garage  BBQ Joint  Baby Store  Bagel Shop    Bakery       Bar  \
0          0.0        0.0         0.0    0.000000  0.052632  0.052632   
1          0.0        0.0         0.0    0.000000  0.100000  0.000000   
2          0.0        0.0         0.0    0.000000  0.032258  0.000000   
3          0.0        0.0         0.0    0.000000  0.052632  0.000000   
4          0.0        0.0         0.0    0.022727  0.000000  0.090909   

   Baseball Field  Beach  Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  \
0             0.0    0.0              0.0       0.0          0.0         0.0   
1             0.0    0.0              0.0       0.0          0.0         0.0   
2             0.0    0.0              0.0       0.0          0.0         0.0   
3             0.0    0.0              0.0       0.0          0.0         0.0   
4             0.0    0.0              0.0       0.0          0.0         0.0   

   Belgian Restaurant  Betting Shop  Bike Rental / Bike Share  Bike Shop  \
0                 0.0           0.0                       0.0        0.0   
1                 0.0           0.0                       0.0        0.0   
2                 0.0           0.0                       0.0        0.0   
3                 0.0           0.0                       0.0        0.0   
4                 0.0           0.0                       0.0        0.0   

     Bistro  Boarding House  Boat or Ferry  Bookstore  Botanical Garden  \
0  0.026316             0.0            0.0   0.026316               0.0   
1  0.000000             0.0            0.0   0.000000               0.0   
2  0.000000             0.0            0.0   0.032258               0.0   
3  0.000000             0.0            0.0   0.052632               0.0   
4  0.022727             0.0            0.0   0.000000               0.0   

   Boutique  Bowling Alle

In [535]:
#Printing each neighborhood along with the 5 most common venues
num_top_venues = 5

for hood in am_lon_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = am_lon_grouped[am_lon_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

---- Bos en Lommer----
                           venue  freq
0                   Tram Station  0.08
1                         Bakery  0.05
2          Indonesian Restaurant  0.05
3                    Coffee Shop  0.05
4  Vegetarian / Vegan Restaurant  0.05


---- Buiksloot----
           venue  freq
0       Bus Stop   0.2
1           Park   0.2
2    Supermarket   0.2
3  Shopping Mall   0.1
4         Bakery   0.1


---- Buikslotermeer----
               venue  freq
0        Supermarket  0.10
1     Clothing Store  0.10
2  Convenience Store  0.06
3          Drugstore  0.06
4     Discount Store  0.03


---- Buitenveldert----
          venue  freq
0     Drugstore  0.11
1   Snack Place  0.05
2  Tram Station  0.05
3   Supermarket  0.05
4        Bakery  0.05


---- De Baarsjes----
                venue  freq
0                 Bar  0.09
1       Deli / Bodega  0.05
2          Restaurant  0.05
3  Falafel Restaurant  0.05
4    Greek Restaurant  0.05


---- De Pijp----
                           ve

In [536]:
#Writing a function to sort the venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [537]:
#Creating a new DF to display the top venues for each neighborhood
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = am_lon_grouped['Neighborhood']

for ind in np.arange(am_lon_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(am_lon_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0    Bos en Lommer          Tram Station                   Bar   
1        Buiksloot           Supermarket              Bus Stop   
2   Buikslotermeer           Supermarket        Clothing Store   
3    Buitenveldert             Drugstore                 Hotel   
4      De Baarsjes                   Bar            Restaurant   

           3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0  Vegetarian / Vegan Restaurant                Bakery  Indonesian Restaurant   
1                           Park         Shopping Mall     Turkish Restaurant   
2                      Drugstore     Convenience Store      Electronics Store   
3           Gym / Fitness Center           Snack Place      French Restaurant   
4             Falafel Restaurant      Greek Restaurant            Supermarket   

  6th Most Common Venue   7th Most Common Venue 8th Most Common Venue  \
0           Coffee Shop               Gastropub           Snack Place   
1            Restaurant                  Bakery     Fish & Chips Shop   
2         Grocery Store                  Bakery        Sandwich Place   
3             Bookstore  Furniture / Home Store           Supermarket   
4    Italian Restaurant                     Pub         Deli / Bodega   

  9th Most Common Venue 10th Most Common Venue  
0       Thai Restaurant          Grocery Store  
1           Fish Market     English Restaurant  
2                 Plaza             Restaurant  
3      Sushi Restaurant           Liquor Store  
4                  Farm                 Bistro

#### Now that our data is in a workable format, we'll start our clustering

In [538]:
# set number of clusters
kclusters = 5

am_lon_grouped_clustering = am_lon_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(am_lon_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [539]:
#creating anew dataframe that includes cluster and the top 10 venues per neighborhood
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [543]:
am_lon_merged.rename(columns = {'neighborhood':'Neighborhood'}, inplace = True)
am_lon_merged.head()

Neighborhood      borough   latitude  longitude    city
0   Abbey Wood       Bexley  51.486042   0.120259  london
1        Acton       Ealing  51.512075  -0.267572  london
2      Aldgate         City  51.515987  -0.078379  london
3      Aldwych  Westminster  51.510583  -0.126768  london
4      Anerley      Bromley  51.412571  -0.061397  london

In [544]:
neighborhoods_venues_sorted.head()

Cluster Labels     Neighborhood 1st Most Common Venue  \
0               0    Bos en Lommer          Tram Station   
1               0        Buiksloot           Supermarket   
2               0   Buikslotermeer           Supermarket   
3               0    Buitenveldert             Drugstore   
4               0      De Baarsjes                   Bar   

  2nd Most Common Venue          3rd Most Common Venue 4th Most Common Venue  \
0                   Bar  Vegetarian / Vegan Restaurant                Bakery   
1              Bus Stop                           Park         Shopping Mall   
2        Clothing Store                      Drugstore     Convenience Store   
3                 Hotel           Gym / Fitness Center           Snack Place   
4            Restaurant             Falafel Restaurant      Greek Restaurant   

   5th Most Common Venue 6th Most Common Venue   7th Most Common Venue  \
0  Indonesian Restaurant           Coffee Shop               Gastropub   
1     Turkish Restaurant            Restaurant                  Bakery   
2      Electronics Store         Grocery Store                  Bakery   
3      French Restaurant             Bookstore  Furniture / Home Store   
4            Supermarket    Italian Restaurant                     Pub   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Snack Place       Thai Restaurant          Grocery Store  
1     Fish & Chips Shop           Fish Market     English Restaurant  
2        Sandwich Place                 Plaza             Restaurant  
3           Supermarket      Sushi Restaurant           Liquor Store  
4         Deli / Bodega                  Farm                 Bistro

In [546]:
am_lon_merged = df_am_lon

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
am_lon_merged = am_lon_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

am_lon_merged 

Neighborhood  \
0                                        Abbey Wood   
1                                             Acton   
2                                           Aldgate   
3                                           Aldwych   
4                                           Anerley   
5                                             Angel   
6                                           Archway   
7                                       Arnos Grove   
8                                            Balham   
9                                          Bankside   
10                                         Barbican   
11                                           Barnes   
12                                        Barnsbury   
13                                        Battersea   
14                                        Bayswater   
15                                     Bedford Park   
16                                        Belgravia   
17                                       Bellingham   
18                                     Belsize Park   
19                                       Bermondsey   
20                                    Bethnal Green   
21                                      Blackfriars   
22                                       Blackheath   
23                        Blackheath Royal Standard   
24                                        Blackwall   
25                                       Bloomsbury   
26                                     Bounds Green   
27                                              Bow   
28                                       Bowes Park   
29                                      Brent Cross   
30                                       Brent Park   
31                                          Brixton   
32                                         Brockley   
33                    Bromley (also Bromley-by-Bow)   
34                                         Brompton   
35                                      Brondesbury   
36                                   Brunswick Park   
37                                   Burroughs, The   
38                                       Camberwell   
39                                  Cambridge Heath   
40                                      Camden Town   
41                                     Canary Wharf   
42                                        Cann Hall   
43                                     Canning Town   
44                                        Canonbury   
45                                        Castelnau   
46                                          Catford   
47                                       Chalk Farm   
48                                    Charing Cross   
49                                         Charlton   
50                                          Chelsea   
51                                      Childs Hill   
52                                        Chinatown   
53                                        Chinbrook   
54                                        Chingford   
55                                         Chiswick   
56                                       Church End   
57                                       Church End   
58                                          Clapham   
59                                      Clerkenwell   
60                                        Colindale   
61                                    Colliers Wood   
62                                     Colney Hatch   
63                                    Covent Garden   
64                                      Cricklewood   
65                                     Crofton Park   
66                                        Crossness   
67                                       Crouch End   
68                                   Crystal Palace   
69                                      Cubitt Town   
70                                     Custom House   
71                                          Dalston   
72                      

In [547]:
am_lon_merged.dropna(inplace = True)

In [548]:
am_lon_merged['Cluster Labels'] = am_lon_merged['Cluster Labels'].astype(int)

In [549]:
# create map for Amsterdam
map_clusters_am = folium.Map(location=[lat_am, lng_am], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(am_lon_merged['latitude'], am_lon_merged['longitude'], am_lon_merged['Neighborhood'], am_lon_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters_am

In [550]:
# create map for London
map_clusters_lon = folium.Map(location=[lat_lon, lng_lon], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(am_lon_merged['latitude'], am_lon_merged['longitude'], am_lon_merged['Neighborhood'], am_lon_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters_lon

#### Examining Clusters

In [556]:
am_lon_merged.columns

Index(['Neighborhood', 'borough', 'latitude', 'longitude', 'city',
       'Cluster Labels', '1st Most Common Venue', '2nd Most Common Venue',
       '3rd Most Common Venue', '4th Most Common Venue',
       '5th Most Common Venue', '6th Most Common Venue',
       '7th Most Common Venue', '8th Most Common Venue',
       '9th Most Common Venue', '10th Most Common Venue'],
      dtype='object')

In [557]:
#Cluster 1
am_lon_merged.loc[am_lon_merged['Cluster Labels'] == 0, am_lon_merged.columns[[1] + [4] + list(range(5, am_lon_merged.shape[1] ))]]

borough       city  Cluster Labels  \
2                                            City     london               0   
3                                     Westminster     london               0   
4                                         Bromley     london               0   
7                                         Enfield     london               0   
11                           Richmond upon Thames     london               0   
14                                    Westminster     london               0   
16                                    Westminster     london               0   
17                                       Lewisham     london               0   
21                                           City     london               0   
24                                  Tower Hamlets     london               0   
25                                         Camden     london               0   
26                                       Haringey     london               0   
28                                       Haringey     london               0   
32                                       Lewisham     london               0   
34   Kensington and ChelseaHammersmith and Fulham     london               0   
36                                         Barnet     london               0   
37                                         Barnet     london               0   
40                                         Camden     london               0   
41                                  Tower Hamlets     london               0   
43                                         Newham     london               0   
45                           Richmond upon Thames     london               0   
46                                       Lewisham     london               0   
47                                         Camden     london               0   
48                                    Westminster     london               0   
49                                      Greenwich     london               0   
50                         Kensington and Chelsea     london               0   
52                                    Westminster     london               0   
56                                          Brent     london               0   
57                                         Barnet     london               0   
62                                         Barnet     london               0   
63                                    Westminster     london               0   
65                                       Lewisham     london               0   
69                                  Tower Hamlets     london               0   
70                                         Newham     london               0   
77                                      Southwark     london               0   
79                         Kensington and Chelsea     london               0   
81                                      Southwark     london               0   
85                                        Enfield     london               0   
91                                       Haringey     london               0   
92                                         Camden     london               0   
95                                       Haringey     london               0   
96                                         Barnet     london               0   
100                                        Barnet     london               0   
104                                     Greenwich     london               0   
113                                        Barnet     london               0   
117                                        Ealing     london               0   
119                                      Haringey     london               0   
120                                        Barnet     london               0   
126                                        Camden     london               0   
127                        Kensington and 

In [558]:
#Cluster 2
am_lon_merged.loc[am_lon_merged['Cluster Labels'] == 1, am_lon_merged.columns[[1] + [4] + list(range(5, am_lon_merged.shape[1]))]]

borough       city  Cluster Labels 1st Most Common Venue  \
30                    Brent     london               1          Liquor Store   
118                   Brent     london               1          Liquor Store   
138                   Brent     london               1          Liquor Store   
157                  Newham     london               1            Restaurant   
162                  Newham     london               1            Restaurant   
191  Hammersmith and Fulham     london               1          Liquor Store   
196                   Brent     london               1          Liquor Store   
244                   Brent     london               1          Liquor Store   
291                   Brent     london               1          Liquor Store   
339                    West  Amsterdam               1            Restaurant   
351               Westpoort  Amsterdam               1                   Bar   

          2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
30               Clothing Store            Restaurant  Fast Food Restaurant   
118              Clothing Store            Restaurant  Fast Food Restaurant   
138              Clothing Store            Restaurant  Fast Food Restaurant   
157         Fried Chicken Joint           Gas Station                   Zoo   
162         Fried Chicken Joint           Gas Station                   Zoo   
191              Clothing Store            Restaurant  Fast Food Restaurant   
196              Clothing Store            Restaurant  Fast Food Restaurant   
244              Clothing Store            Restaurant  Fast Food Restaurant   
291              Clothing Store            Restaurant  Fast Food Restaurant   
339  Modern European Restaurant         Boat or Ferry                   Zoo   
351                  Restaurant                 Field    English Restaurant   

     5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
30      English Restaurant  Entertainment Service  Ethiopian Restaurant   
118     English Restaurant  Entertainment Service  Ethiopian Restaurant   
138     English Restaurant  Entertainment Service  Ethiopian Restaurant   
157   Fast Food Restaurant  Entertainment Service  Ethiopian Restaurant   
162   Fast Food Restaurant  Entertainment Service  Ethiopian Restaurant   
191     English Restaurant  Entertainment Service  Ethiopian Restaurant   
196     English Restaurant  Entertainment Service  Ethiopian Restaurant   
244     English Restaurant  Entertainment Service  Ethiopian Restaurant   
291     English Restaurant  Entertainment Service  Ethiopian Restaurant   
339    Filipino Restaurant   Ethiopian Restaurant           Event Space   
351  Entertainment Service   Ethiopian Restaurant           Event Space   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
30            Event Space               Exhibit     Falafel Restaurant  
118           Event Space               Exhibit     Falafel Restaurant  
138           Event Space               Exhibit     Falafel Restaurant  
157           Event Space               Exhibit     Falafel Restaurant  
162           Event Space               Exhibit     Falafel Restaurant  
191           Event Space               Exhibit     Falafel Restaurant  
196           Event Space               Exhibit     Falafel Restaurant  
244           Event Space               Exhibit     Falafel Restaurant  
291           Event Space               Exhibit     Falafel Restaurant  
339               Exhibit    Falafel Restaurant                   Farm  
351               Exhibit    Falafel Restaurant                   Farm

In [562]:
#Cluster 3
am_lon_merged.loc[am_lon_merged['Cluster Labels'] == 2, am_lon_merged.columns[[1] + [4] + list(range(5, am_lon_merged.shape[1]))]]

borough    city  Cluster Labels 1st Most Common Venue  \
0                    Bexley  london               2                Bakery   
1                    Ealing  london               2         Grocery Store   
5                 Islington  london               2                   Pub   
6                 Islington  london               2           Coffee Shop   
8                Wandsworth  london               2           Coffee Shop   
9                 Southwark  london               2                   Pub   
10                     City  london               2                   Pub   
12                Islington  london               2                   Pub   
13               Wandsworth  london               2                   Pub   
15                   Ealing  london               2                  Café   
18                   Camden  london               2                  Café   
19                Southwark  london               2                   Pub   
20            Tower Hamlets  london               2           Coffee Shop   
27            Tower Hamlets  london               2                   Pub   
29                   Barnet  london               2                   Pub   
31                  Lambeth  london               2              Bus Stop   
33            Tower Hamlets  london               2                   Pub   
35                    Brent  london               2                  Café   
38                Southwark  london               2                  Café   
39            Tower Hamlets  london               2           Coffee Shop   
42           Waltham Forest  london               2                   Pub   
44                Islington  london               2                   Pub   
51                   Barnet  london               2                   Pub   
53                 Lewisham  london               2         Grocery Store   
54           Waltham Forest  london               2                   Pub   
55                 Hounslow  london               2                  Café   
58                  Lambeth  london               2                   Pub   
59                Islington  london               2                   Pub   
60                   Barnet  london               2                   Pub   
61                   Merton  london               2                   Pub   
64                   Barnet  london               2                   Pub   
66                   Bexley  london               2                Bakery   
67                 Haringey  london               2                  Café   
68                  Bromley  london               2           Coffee Shop   
71                  Hackney  london               2                   Pub   
72                 Dartford  london               2                   Pub   
73                Islington  london               2                   Pub   
74                Southwark  london               2                  Café   
75                 Lewisham  london               2                   Pub   
76                    Brent  london               2                   Pub   
78                   Ealing  london               2                   Pub   
80               Wandsworth  london               2                   Pub   
82                   Barnet  london               2           Coffee Shop   
83                   Newham  london               2         Grocery Store   
84     Richmond upon Thames  london               2           Coffee Shop   
86                Southwark  london               2                   Pub   
88         Islington & City  london               2                   Pub   
89                   Barnet  london               2           Coffee Shop   
90                Islington  london               2                   Pub   
93                   Newham  london               2                   Pub   
94                 Lewisham  london               2         Grocery Store   
97              

In [561]:
#Cluster 4
am_lon_merged.loc[am_lon_merged['Cluster Labels'] == 3, am_lon_merged.columns[[1] + [4] + list(range(5, am_lon_merged.shape[1]))]]

borough    city  Cluster Labels 1st Most Common Venue  \
22               Lewisham  london               3                  Lake   
23              Greenwich  london               3                  Lake   
141             Greenwich  london               3                  Lake   
145  Kingston upon Thames  london               3      Business Service   
206            Wandsworth  london               3      Business Service   
210            Wandsworth  london               3      Business Service   
286             Greenwich  london               3                  Lake   

          2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22             Business Service                   Zoo                 Field   
23             Business Service                   Zoo                 Field   
141            Business Service                   Zoo                 Field   
145  Construction & Landscaping     Fish & Chips Shop  Ethiopian Restaurant   
206  Construction & Landscaping     Fish & Chips Shop  Ethiopian Restaurant   
210  Construction & Landscaping     Fish & Chips Shop  Ethiopian Restaurant   
286            Business Service                   Zoo                 Field   

     5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22   Entertainment Service  Ethiopian Restaurant           Event Space   
23   Entertainment Service  Ethiopian Restaurant           Event Space   
141  Entertainment Service  Ethiopian Restaurant           Event Space   
145            Event Space               Exhibit    Falafel Restaurant   
206            Event Space               Exhibit    Falafel Restaurant   
210            Event Space               Exhibit    Falafel Restaurant   
286  Entertainment Service  Ethiopian Restaurant           Event Space   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22                Exhibit    Falafel Restaurant                   Farm  
23                Exhibit    Falafel Restaurant                   Farm  
141               Exhibit    Falafel Restaurant                   Farm  
145                  Farm        Farmers Market   Fast Food Restaurant  
206                  Farm        Farmers Market   Fast Food Restaurant  
210                  Farm        Farmers Market   Fast Food Restaurant  
286               Exhibit    Falafel Restaurant                   Farm

In [560]:
#Cluster 5
am_lon_merged.loc[am_lon_merged['Cluster Labels'] == 4, am_lon_merged.columns[[1] + [4] + list(range(5, am_lon_merged.shape[1]))]]

borough    city  Cluster Labels 1st Most Common Venue  \
87   Greenwich  london               4                  Park   
168  Greenwich  london               4                  Park   
174    Bromley  london               4                  Park   
179  Greenwich  london               4                  Park   
275  Greenwich  london               4                  Park   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
87            Golf Course          Soccer Field                   Zoo   
168           Golf Course          Soccer Field                   Zoo   
174           Golf Course          Soccer Field                   Zoo   
179           Golf Course          Soccer Field                   Zoo   
275           Golf Course          Soccer Field                   Zoo   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
87     English Restaurant  Ethiopian Restaurant           Event Space   
168    English Restaurant  Ethiopian Restaurant           Event Space   
174    English Restaurant  Ethiopian Restaurant           Event Space   
179    English Restaurant  Ethiopian Restaurant           Event Space   
275    English Restaurant  Ethiopian Restaurant           Event Space   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
87                Exhibit    Falafel Restaurant                   Farm  
168               Exhibit    Falafel Restaurant                   Farm  
174               Exhibit    Falafel Restaurant                   Farm  
179               Exhibit    Falafel Restaurant                   Farm  
275               Exhibit    Falafel Restaurant                   Farm

### Looking at the clusters, we have many the most homogeneity in cluster 1 meaning that those boroughs are probably the most similar. If you are looking for something a bit different, it seems that Boroughs in cluster 2-5 would be something to look into.